packages/lodestar/src/network/encoders/request.ts

```diff
@@ -53,76 +53,123 @@ export function eth2RequestDecode(
+      sszDataLength = decode(chunk.slice());
+      if (decode.bytes > 10) {
+        throw new RequestDecodeError({code: RequestDecodeErrorCode.INVALID_VARINT_BYTES_COUNT, bytes: decode.bytes});
       }
 
-      if (sszDataLength < type.minSize() || sszDataLength > type.maxSize()) {
-        logger.error("eth2RequestDecode: Invalid szzLength", {sszDataLength, method});
-        yield {isValid: false};
-        break;
+      chunk = chunk.slice(decode.bytes);
+      if (chunk.length === 0) {
+        continue;
       }
+    }
 
-      if (chunk.length > maxEncodedLen(sszDataLength, encoding)) {
-        logger.error("eth2RequestDecode: too much bytes read", {chunkLength: chunk.length, sszDataLength, method});
-        yield {isValid: false};
-        break;
-      }
+    if (sszDataLength < minSize) {
+      throw new RequestDecodeError({code: RequestDecodeErrorCode.UNDER_SSZ_MIN_SIZE, minSize, sszDataLength});
+    }
+    if (sszDataLength > maxSize) {
+      throw new RequestDecodeError({code: RequestDecodeErrorCode.OVER_SSZ_MAX_SIZE, maxSize, sszDataLength});
+    }
 
-      let uncompressed: Buffer | null = null;
-      try {
-        uncompressed = decompressor.uncompress(chunk.slice());
-      } catch (e) {
-        logger.error("Failed to decompress request data", {error: e.message});
-        yield {isValid: false};
-        break;
-      }
+    const chunkLength = chunk.length;
+    if (chunkLength > maxEncodedLen(sszDataLength, encoding)) {
+      throw new RequestDecodeError({code: RequestDecodeErrorCode.TOO_MUCH_BYTES_READ, chunkLength, sszDataLength});
+    }
 
-      if (uncompressed) {
-        buffer.append(uncompressed);
-      }
+    let uncompressed: Buffer | null = null;
+    try {
+      uncompressed = decompressor.uncompress(chunk.slice());
+    } catch (e) {
+      throw new RequestDecodeError({code: RequestDecodeErrorCode.DECOMPRESSOR_ERROR, decompressorError: e});
+    }
 
-      if (buffer.length < sszDataLength) {
-        continue;
-      }
+    if (uncompressed) {
+      buffer.append(uncompressed);
+    }
 
-      if (buffer.length > sszDataLength) {
-        logger.error("Too long message received", {method});
-        yield {isValid: false};
-        break;
-      }
+    if (buffer.length < sszDataLength) {
+      continue;
+    }
 
-      try {
-        yield {isValid: true, body: type.deserialize(buffer.slice(0, sszDataLength)) as RequestBody};
-      } catch (e) {
-        logger.error("Malformed input. Failed to deserialize request type", {method, error: e.message});
-        yield {isValid: false};
-        break;
-      }
-      // only one request body accepted
-      break;
+    if (buffer.length > sszDataLength) {
+      throw new RequestDecodeError({code: RequestDecodeErrorCode.TOO_MANY_BYTES, sszDataLength});
     }
-  };
+
+    // buffer.length === sszDataLength
+
+    // only one request body accepted, so return
+    try {
+      return type.deserialize(buffer.slice(0, sszDataLength));
```

Comment:
I wonder if you don't break when reading from source, do you even close stream. It's been a while,
but I remember that it could hang open if you don't read everything or break. libp2p handles open and closes in post loop cleanup. I know we had some issues because if you don't finish reading that post iteration cleanup did not occur.